akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -1048,38 +1049,39 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
     private case object Initializing extends State
     // stream is initialized and so no threads can just send events without any synchronization overhead
     private case object Initialized extends State
-    // stage has been shut down, either regularly or it failed
-    private case object Completed extends State
     // Event with feedback promise
     private case class Event(e: T, handlingPromise: OptionVal[Promise[Done]])
 
-    val waitingForProcessing = ConcurrentHashMap.newKeySet[Promise[_]]()
-
     private[this] val currentState = new AtomicReference[State](Pending(Nil))
 
     // is called from the owning [[GraphStage]]
     @tailrec
-    final private[stage] def onStart(): Unit = {
+    private[stage] def onStart(): Unit = {
       (currentState.getAndSet(Initializing): @unchecked) match {
         case Pending(l) ⇒ l.reverse.foreach(evt ⇒ {
+          evt.handlingPromise match {
+            case OptionVal.Some(p) ⇒ p.future.onComplete(_ ⇒ onFeedbackCompleted(p))(ExecutionContexts.sameThreadExecutionContext)
+            case OptionVal.None    ⇒
+          }
           onAsyncInput(evt.e, evt.handlingPromise)
         })
       }
       if (!currentState.compareAndSet(Initializing, Initialized)) {
         (currentState.get: @unchecked) match {
           case Pending(_) ⇒ onStart()
-          case Completed  ⇒ () //wonder if this is possible
         }
       }
     }
 
     // is called from the owning [[GraphStage]]
-    final private[stage] def onStop(): Unit = {
-      currentState.set(Completed)
-      val iterator = waitingForProcessing.iterator()
-      lazy val detachedException = new StreamDetachedException()
-      while (iterator.hasNext) { iterator.next().tryFailure(detachedException) }
-      waitingForProcessing.clear()
+    private[stage] def onStop(outstandingPromises: Set[Promise[Done]]): Unit = {
+      if (outstandingPromises.nonEmpty) {
+        lazy val detachedException = new StreamDetachedException()
```

Comment:
Maybe less important than the starting part, but still the lazy val imposes volatile ops, would we prefer to avoid those?

Reply:
Yeah, this was old code that ended up here, we don't need the lazy, there is already a guard against empty for the entire block where the exception is used.